jetty-core/jetty-quic/jetty-quic-server/src/main/java/org/eclipse/jetty/quic/server/ServerQuicConnection.java

```diff
@@ -114,6 +116,20 @@ protected QuicSession createSession(SocketAddress remoteAddress, ByteBuffer ciph
         }
         else
         {
+            if (quicConfiguration.getSslContextFactory().getNeedClientAuth())
+            {
+                byte[] peerCertificate = quicheConnection.getPeerCertificate();
+                if (peerCertificate == null)
+                {
+                    ServerQuicSession session = newQuicSession(remoteAddress, quicheConnection);
+                    ProtocolSession protocolSession = session.createProtocolSession();
+                    protocolSession.disconnect(QuicErrorCode.CONNECTION_REFUSED.code(), "missing_client_cert");
+                    // Send the response packet(s) that disconnect() generated.
+                    session.flush();
+                    return null;
+                }
+            }
+
             ServerQuicSession session = newQuicSession(remoteAddress, quicheConnection);
             // Send the response packet(s) that tryAccept() generated.
             session.flush();
```

Comment:
```suggestion
            ServerQuicSession session = newQuicSession(remoteAddress, quicheConnection);

            if (quicConfiguration.getSslContextFactory().getNeedClientAuth())
            {
                byte[] peerCertificate = quicheConnection.getPeerCertificate();
                if (peerCertificate == null)
                {
                    session.outwardClose(QuicErrorCode.CONNECTION_REFUSED.code(), "missing_client_certificate");
                    return null;
                }
            }

            // Send the response packet(s) that tryAccept() generated.
            session.flush();
            return session;
```

There should be no need to create a `ProtocolSession`, to call `disconnect()` on it, just to have `session.outwardClose()` be called.

These are the initial steps of connection establishment, so I think we should just close at the QUIC level, exactly as if some other TLS validation failed -- no need to create the upper layers.

If that is correct, I'd remove the added `disconnect()` methods and overrides -- it should be much simpler.

Reply:
The `disconnect()` call isn't just to call `session.outwardClose()` but also to call `onDisconnect()` on the session's listener, see the `ServerHTTP3Session.disconnect()` impl.